page-tracking/src/page_tracker.rs

```diff
@@ -274,8 +286,22 @@ impl PageTracker {
         Ok(unsafe { P::DirtyPage::new(addr) })
     }
 
-    /// Releases an exclusive reference to a converted and locked page.
-    pub fn put_converted_page<P: ConvertedPhysPage>(&self, page: P) -> Result<()> {
+    /// Returns the lock status of the page
+    pub fn is_locked_page(&self, addr: SupervisorPageAddr) -> Result<bool> {
+        let mut page_tracker = self.inner.lock();
+        let info = page_tracker.get_mut(addr)?;
+        Ok(info.is_locked())
+    }
+
+    /// Obtains an exclusive reference on an assignable page
+    pub fn lock_for_assignment(&self, addr: SupervisorPageAddr) -> Result<()> {
```

Comment:
Why do we need public APIs to lock a page or check if a page is locked? There's only two cases where we need to lock a page presently: when acquiring a reference to a converted page (for assignment/reclaim) or acquiring a reference to a mapped page (for sharing); can we keep the locking internal to `get_converted_page()` and `get_shareable_page()`?

Also, see my API below on trying to avoid `PageTracker` APIs that operate on raw addresses.

Reply:
It all came down to the same underlying cause -- pages allocated by HyperAlloc() were in Converted state, but not locked. Adding a lock() there obviates the need for public APIs to check for and lock pages in the tracker.